react/react_test.go

```diff
@@ -21,47 +24,96 @@ func TestTestVersion(t *testing.T) {
-	}
+	assertCellValue(t, i, 2, "i.Value() doesn't match changed value")
 }
 
-// The value of a compute cell is determined by the value of the dependencies.
-func TestBasicCompute(t *testing.T) {
+// The value of a compute 1 cell is determined by the value of the dependencies.
+func TestBasicCompute1(t *testing.T) {
 	r := New()
 	i := r.CreateInput(1)
 	c := r.CreateCompute1(i, func(v int) int { return v + 1 })
-	if c.Value() != 2 {
-		t.Fatalf("c.Value() isn't properly computed based on initial input cell value")
-	}
+	assertCellValue(t, c, 2, "c.Value() isn't properly computed based on initial input cell value")
 	i.SetValue(2)
-	if c.Value() != 3 {
-		t.Fatalf("c.Value() isn't properly computed based on changed input cell value")
-	}
+	assertCellValue(t, c, 3, "c.Value() isn't properly computed based on changed input cell value")
+}
+
+// The value of a compute 2 cell is determined by the value of the dependencies.
+func TestBasicCompute2(t *testing.T) {
+	r := New()
+	i1 := r.CreateInput(1)
+	i2 := r.CreateInput(2)
+	c := r.CreateCompute2(i1, i2, func(v1, v2 int) int { return v1 | v2 })
+	assertCellValue(t, c, 3, "c.Value() isn't properly computed based on initial input cell values")
+	i1.SetValue(4)
+	assertCellValue(t, c, 6, "c.Value() isn't properly computed when first input cell value changes")
+	i2.SetValue(8)
+	assertCellValue(t, c, 12, "c.Value() isn't properly computed when second input cell value changes")
 }
 
-// Compute cells can depend on other compute cells.
+// Compute 2 cells can depend on compute 1 cells.
 func TestTopology(t *testing.T) {
 	r := New()
 	i := r.CreateInput(1)
 	c1 := r.CreateCompute1(i, func(v int) int { return v + 1 })
 	c2 := r.CreateCompute1(i, func(v int) int { return v - 1 })
 	c3 := r.CreateCompute2(c1, c2, func(v1, v2 int) int { return v1 * v2 })
-	if c3.Value() != 0 {
-		t.Fatalf("c3.Value() isn't properly computed based on initial input cell value")
-	}
+	assertCellValue(t, c3, 0, "c3.Value() isn't properly computed based on initial input cell value")
 	i.SetValue(3)
-	if c3.Value() != 8 {
-		t.Fatalf("c3.Value() isn't properly computed based on changed input cell value")
+	assertCellValue(t, c3, 8, "c3.Value() isn't properly computed based on changed input cell value")
+}
+
+// Compute 1 cells can depend on other compute 1 cells.
+func TestCompute1Chain(t *testing.T) {
+	r := New()
+	inp := r.CreateInput(1)
+	var c Cell = inp
+	for i := 2; i <= 8; i++ {
+		// must save current value of loop variable i for correct behavior.
+		// compute function has to use digitToAdd not i.
+		digitToAdd := i
+		c = r.CreateCompute1(c, func(v int) int { return v*10 + digitToAdd })
+	}
+	assertCellValue(t, c, 12345678, "c.Value() isn't properly computed based on initial input cell value")
+	inp.SetValue(9)
+	assertCellValue(t, c, 92345678, "c.Value() isn't properly computed based on changed input cell value")
+}
+
+// Compute 2 cells can depend on other compute 2 cells.
+func TestCompute2Tree(t *testing.T) {
```

Comment:
I'll hold off. This test already covers that compute2 can depend on compute2 for one level of nesting, and I've tested many levels of nesting in the compute1 chain. I can't imagine someone would make a solution that passes both those tests but that wouldn't support trees with many levels, so I suspect a large tree test would add little value.